src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -428,55 +432,50 @@ def execute(self):
-    # all targets to finish.
-    with self.invalidated(relevant_targets,
-                          invalidate_dependents=True,
-                          fingerprint_strategy=fingerprint_strategy,
-                          topological_order=True) as invalidation_check:
-
-      compile_contexts = {vt.target: self.create_compile_context(vt.target, vt.results_dir)
-                          for vt in invalidation_check.all_vts}
-
-      self.do_compile(
-        invalidation_check,
-        compile_contexts,
-        classpath_product,
-      )
 
-      if not self.get_options().use_classpath_jars:
-        # Once compilation has completed, replace the classpath entry for each target with
-        # its jar'd representation.
-        for ccs in compile_contexts.values():
-          cc = self.select_runtime_context(ccs)
-          for conf in self._confs:
-            classpath_product.remove_for_target(cc.target, [(conf, cc.classes_dir)])
-            classpath_product.add_for_target(cc.target, [(conf, cc.jar_file)])
-
-      # The runtime classpath will always be a superset of the export_dep_as_jar_classpath
-      # so however we compute runtime_classpath it will contain all the nescessary jars for
-      # the export_dep_as_jar_classpath.export_dep_as_jar_classpath.
-      if self.context.products.is_required_data('export_dep_as_jar_classpath'):
-        self.context.products.get_data('export_dep_as_jar_classpath', classpath_product.copy)
+    dependees_of_target_roots = None
+    # If we are only exporting jars then we can omit some targets from the runtime_classpath.
+    if self.context.products.is_required_data("export_dep_as_jar_signal"):
+      # Filter modulized targets from invalid targets list.
+      target_roots_in_play = set(relevant_targets) & set(self.context.target_roots)
+      addresses_in_play = [t.address for t in target_roots_in_play]
+      dependees_of_target_roots = self.context.build_graph.transitive_dependees_of_addresses(addresses_in_play)
+      relevant_targets = list(set(relevant_targets) - dependees_of_target_roots)
+
+    if relevant_targets:
+      # Note, JVM targets are validated (`vts.update()`) as they succeed.  As a result,
+      # we begin writing artifacts out to the cache immediately instead of waiting for
+      # all targets to finish.
+      with self.invalidated(relevant_targets,
+                            invalidate_dependents=True,
+                            fingerprint_strategy=fingerprint_strategy,
+                            topological_order=True) as invalidation_check:
+
+        compile_contexts = {vt.target: self.create_compile_context(vt.target, vt.results_dir)
+                            for vt in invalidation_check.all_vts}
+
+        self.do_compile(
+          invalidation_check,
+          compile_contexts,
+          classpath_product,
+        )
+
+        if not self.get_options().use_classpath_jars:
```

Comment:
I didn't modify this line, but the conditional looks fishy to me. Seems like we should be using `if self.get_options().use_classpath_jars` based on what the block does. Does anyone have context on this?

Reply:
Unknown. The last real change on this line was: 93894d4d59a5d80afb3dfc7be0a7d5a49d497b29, so likely no need to touch it.

Reply:
Not going to, but it is strange.